### PR TITLE
Reduce overhead data spec colours

### PR DESCRIPTION
**Pull request by rowleya**
Once #1196 is merged, this can follow.

Adds "colours" to allow processing of spikes from previous time steps provided the delays are big enough to cope with the difference.

Depends on:
 - SpiNNakerManchester/PACMAN#480
 - SpiNNakerManchester/SpiNNFrontEndCommon#1007

Tested by SpiNNakerManchester/IntegrationTests#168

### CODE REVIEW

File: neural_modelling/src/neuron/synapses.c [hunk 1]
```diff
@@ -80,9 +80,21 @@ uint32_t synapse_delay_mask;
...
+//! Count of the spikes that are received late
+uint32_t late_spikes = 0;
+
+//! The maximum lateness of a spike
+uint32_t max_late_spike = 0;
+
 //! Number of neurons
 static uint32_t n_neurons_peak;
 
+static uint32_t synapse_delay_mask_shifted = 0;
```
Comment (andrewgait) [inline neural_modelling/src/neuron/synapses.c:96]: Missing doc string

File: neural_modelling/src/neuron/synapses.c [hunk 2]
```diff
@@ -328,7 +349,25 @@ void synapses_flush_ring_buffers(timer_t time) {
 }
 
 bool synapses_process_synaptic_row(
-        uint32_t time, synaptic_row_t row, bool *write_back) {
+        uint32_t time, uint32_t spike_colour, uint32_t colour_mask,
+		synaptic_row_t row, bool *write_back) {
+
+    // Work out how much delay take off or add on to the actual delay because
```
Comment (andrewgait) [inline neural_modelling/src/neuron/synapses.c:355]: "takes off or adds on"

File: neural_modelling/src/neuron/synapses.c [hunk 3]
```diff
@@ -328,7 +349,25 @@ void synapses_flush_ring_buffers(timer_t time) {
...
+    // of a delayed spike arrival time, or delayed change of time step in the
+    // current core.  Spikes can be as late as the bits in colour_mask dictates.
+	/// Masked difference is used to calculate this, which will always be
```
Comment (andrewgait) [inline neural_modelling/src/neuron/synapses.c:358]: Looks like a stray / at the start of this line

File: neural_modelling/src/neuron/synapses.c [hunk 4]
```diff
@@ -328,7 +349,25 @@ void synapses_flush_ring_buffers(timer_t time) {
 }
 
 bool synapses_process_synaptic_row(
-        uint32_t time, synaptic_row_t row, bool *write_back) {
+        uint32_t time, uint32_t spike_colour, uint32_t colour_mask,
+		synaptic_row_t row, bool *write_back) {
+
+    // Work out how much delay take off or add on to the actual delay because
+    // of a delayed spike arrival time, or delayed change of time step in the
+    // current core.  Spikes can be as late as the bits in colour_mask dictates.
```
Comment (andrewgait) [inline neural_modelling/src/neuron/synapses.c:357]: I suppose as well that you can't indicate too many bits in colour_mask due to other limits on delays?

File: spynnaker/pyNN/models/neuron/abstract_population_vertex.py [hunk 5]
```diff
@@ -157,14 +158,15 @@ class AbstractPopulationVertex(
     _SYNAPSE_BASE_N_CPU_CYCLES_PER_NEURON = 22
     _SYNAPSE_BASE_N_CPU_CYCLES = 10
 
-    # 5 elements before the start of global parameters
-    # 1. has key, 2. key, 3. n atoms, 4. n_atoms_peak 5. n_synapse_types
+    # Elements before the start of global parameters
+    # 1. has key, 2. key, 3. n atoms, 4. n_atoms_peak 5. n_colour_bits
```
Comment (andrewgait) [inline spynnaker/pyNN/models/neuron/abstract_population_vertex.py:162]: This change looked a bit strange to start off with until I realised that n_synapse_types wasn't in this list of elements any more (perhaps it used to be... ?)

File: spynnaker/pyNN/models/neuron/population_machine_synapses_provenance.py [hunk 6]
```diff
@@ -129,3 +138,17 @@ def _parse_synapse_provenance(self, label,  x, y, p, provenance_data):
...
+            if synapse_prov.n_skipped_synapses > 0:
+                db.insert_report(
+                    f"On {label}, there were {synapse_prov.n_skipped_synapses}"
+                    " that were skipped because a spike was received later"
+                    " than the delay in the synapse")
+
+            db.insert_core(
+                x, y, p, self.LATE_SPIKES, synapse_prov.n_late_spikes)
+            db.insert_core(
+                x, y, p, self.MAX_LATE_SPIKE, synapse_prov.max_late_spike)
```
Comment (andrewgait) [inline spynnaker/pyNN/models/neuron/population_machine_synapses_provenance.py:154]: Perhaps these (n_late_spikes and max_late_spike) should also go in the report?

File: spynnaker/pyNN/models/utility_models/delays/delay_extension_machine_vertex.py [hunk 7]
```diff
@@ -230,7 +231,7 @@ def generate_data_specification(self, spec, placement):
         # Reserve memory:
         spec.comment("\nReserving memory space for data regions:\n\n")
 
-        # ###################################################################
+        # ###################################################################self
```
Comment (andrewgait) [inline spynnaker/pyNN/models/utility_models/delays/delay_extension_machine_vertex.py:234]: Looks like a "self" has been mistakenly added to the end of this line

**Review by andrewgait — CHANGES_REQUESTED** [review]
This generally looks fine, just a few questions/comments.

File: neural_modelling/src/neuron/synapses.c (hunk 1 above)
Comment (rowleya) [reply neural_modelling/src/neuron/synapses.c:97]: Added

File: neural_modelling/src/neuron/synapses.c (hunk 2 above)
Comment (rowleya) [reply neural_modelling/src/neuron/synapses.c]: Fixed

File: neural_modelling/src/neuron/synapses.c (hunk 3 above)
Comment (rowleya) [reply neural_modelling/src/neuron/synapses.c]: Fixed

File: spynnaker/pyNN/models/neuron/abstract_population_vertex.py (hunk 5 above)
Comment (rowleya) [reply spynnaker/pyNN/models/neuron/abstract_population_vertex.py:162]: Yes, it was already a comment error; I checked that the correct number of elements was now added and just changed the mistake!

File: spynnaker/pyNN/models/utility_models/delays/delay_extension_machine_vertex.py (hunk 7 above)
Comment (rowleya) [reply spynnaker/pyNN/models/utility_models/delays/delay_extension_machine_vertex.py]: Oops - removed!

**Review by andrewgait — APPROVED** [review]

**Pull request merged**